### PR TITLE
[Feature] Feature Overview for Exam Mode

### PR DESCRIPTION
**Pull request by anditurdiu**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
We want to create a page which displays the features of the exam mode in a clear way.
We also made some changes to the exam end page to reduce the clutter. 

### Description
- Create the general structure of the view. 
- Create different routes for students and instructors. 
- Lazy load the features overview module
- Create a pre-configured structure for texts including 1 text, 1 image, 2 texts, 2 images, centre text, first image left, first image right
- Add the features displayed in: 

> https://confluence.ase.in.tum.de/display/ArTEMiS/Feature+Overview+German
> https://confluence.ase.in.tum.de/display/ArTEMiS/Feature+Overview+English

- Add the text for each feature
- Add the screenshots where relevant

### Coverage
feature-overview.component.ts 67.5%

### Steps for Testing
1. nagivate to /#/features/instructors
2. navigate to /#/features/students
When you hover over a feature, more information is displayed.
When you press on a feature you should automatically scroll to the feature details.

### Screenshots
image
image
image

### CODE REVIEW

**Review by kloessst — CHANGES_REQUESTED** [review]
I like the layout and the animation looks cool. Just found a small issue when the short text is displayed again. The line wrapping changes because the text gets bigger. This creates some visual noise.
wrapIssue

In the student view:
- Image for intuitive user interface, user your own IDE are a bit blurry blurry
- LRZ and TUM logo is also a bit blurry

Generally I think some images are too small to really see what's going on e.g. Integrated Anonymous Assessment on instructors view. Maybe a fixed image size would help, or make the images expandable by clicking on them.

Can you only access those pages using the links? Would be nice to have this linked on the landing page in the future.

**Review by kokowaahmax — CHANGES_REQUESTED** [review]
Well work and really nice done! Only some small suggestions that I think you can make some changes:

------
Screenshot 2020-10-13 at 11 49 44 Screenshot 2020-10-13 at 11 49 34 Screenshot 2020-10-13 at 12 55 22

This three cube have different format than other cubes, normally it starts with a verb to discribe an action like "Add a variety of exercises..." But only in these three are in different format

------
Screenshot 2020-10-13 at 11 54 03

There are empty space between dividing line, do you want to add an icon inside? If not you can just delete the white space

------

Screenshot 2020-10-13 at 12 01 58

It says the exam mode support file-up load exercise, but the screenshot actually doesn't support

------

Screenshot 2020-10-13 at 12 06 01

Probably change the alignment from center to leading here would be better, and an entire line with only one word should be avoid

------
Screenshot 2020-10-13 at 12 03 53
between sth or sth / between sth and sth?
doing sth or can do sth sounds weird.

Screenshot 2020-10-13 at 12 13 08
generated?

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]
Have some changes for text style and spelling, discussing this directly over Zoom.

**Comment by anditurdiu** [comment]
Thank you, everyone, for the review. I added the new version of the texts after discussing it with @MaisiKoleni.
The texts should be free of typos now and better worded than before. If you want to suggest more changes, it is easier to change it here: https://confluence.ase.in.tum.de/display/ArTEMiS/Feature+Overview+Page 
using the `Page History` function I can directly see what improvement suggestions you have. This is easier for all of us instead of trying to read through the JSON :D

**Comment by anditurdiu** [comment]
> Can you only access those pages using the links? Would be nice to have this linked on the landing page in the future.
**Visual noise:** 
I do not really know a way around the visual noise when you hover over it. What is happening is that the view is shrunk down to 50%, which is why the text also changes. But I think it is fine for now.

**With regards to the image resolutions:**
I will have access tomorrow to a 4k screen so I can take higher resolution screenshots. I'm hoping that will solve the issue. I have set a fixed maximum width of 800px per image. But for some reason it is forcing lower resolution images to scale up which is causing the noise. 

**Access to the feature overview:**
For now it is intended that you can only access it through the link. Maybe in the future we can add it to the initial screen :)

**Comment by krusche** [comment]
The exam mode does not support file upload exercises. Please take this into account in the corresponding texts

**Comment by anditurdiu** [comment]
> Well work and really nice done! Only some small suggestions that I think you can make some changes:
> 
> Screenshot 2020-10-13 at 11 49 44 Screenshot 2020-10-13 at 11 49 34 Screenshot 2020-10-13 at 12 55 22
> 
> This three cube have different format than other cubes, normally it starts with a verb to discribe an action like "Add a variety of exercises..." But only in these three are in different format
> 
> Screenshot 2020-10-13 at 11 54 03
> 
> There are empty space between dividing line, do you want to add an icon inside? If not you can just delete the white space
> 
> Screenshot 2020-10-13 at 12 01 58
> 
> It says the exam mode support file-up load exercise, but the screenshot actually doesn't support
> 
> Screenshot 2020-10-13 at 12 06 01
> 
> Probably change the alignment from center to leading here would be better, and an entire line with only one word should be avoid
> 
> Screenshot 2020-10-13 at 12 03 53
> between sth or sth / between sth and sth?
> doing sth or can do sth sounds weird.
> 
> Screenshot 2020-10-13 at 12 13 08
> generated?

**title not starting with a verb:**
If you have any suggestion for a different wording I'm happy to change it :) 

**Divider whitespace:**
That was a styling choice on my side, the undivided divider looked strange to me. 

**Alignment**
I can change the center alignment, but I'm not sure you can set a minimum amount of words per line. If you change the width of the window you will always be able to reach a point where a word is pushed to the next line.

**Comment by anditurdiu** [comment]
> The exam mode does not support file upload exercises. Please take this into account in the corresponding texts

I removed the file upload mentions

**Comment by anditurdiu** [comment]
I added 4k resolution images to replace the old screenshots. I also annotated some of them and used the .gif format for screens which were too long. The images are still displayed as slightly blurry but I think this is due to the resizing effect of .png's which contain fine content such as text.
Is it good enough as it is? Otherwise, I could implement a resizing feature as a follow up where the items expand to full width once a user clicks on them. They could appear as too large though due to the new resolution.

**Review by tiffca — APPROVED** [review]
All looks good, tested on test server 1 on chrome, firefox, safari. 
Read through the confluence page (English) for the wording, nothing major noticed, nice job! 

minor note: I also noticed the discrepancy between "All known Artemis exercise types are supported within the exam mode" and the reality that file upload exercises are not supported. 

Overall it looks really good and functions well!

**Comment by kloessst** [comment]
> 
> 
> I added 4k resolution images to replace the old screenshots and annotated. The images are still displayed as slightly blurry but I think this is due to the resizing effect of .png's which contain fine content such as text.
> Is it good enough as it is? Otherwise, I could implement a resizing feature as a follow up where the items expand to full width once a user clicks on them. They could appear as too large though due to the new resolution.
> 
> -Content is also up to date with the latest version on confluence for german and english-

I think this is good for the first iteration. Making the the pictures expandable would be nice in the future. I also think showing some short gifs or videos of the UI would be nice.

**Review by kloessst — DISMISSED** [review]

File: src/main/webapp/app/feature-overview/feature-overview.component.ts [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+
+    private setupStudentFeatures() {
+        // set up student features
+        const featurelocalIDE = new Feature(
+            'featureOverview.students.feature.localIDE.title',
+            'featureOverview.students.feature.localIDE.shortDescription',
+            'featureOverview.students.feature.localIDE.descriptionTextOne',
+            'fa fa-code-fork',
+            undefined,
+            'https://github.com/ls1intum/Artemis/blob/feature/informative-marketing/src/main/resources/public/images/feature-overview/students/clone_repository.png?raw=true',
```
Comment (krusche) [inline src/main/webapp/app/feature-overview/feature-overview.component.ts:39]: please load those images from the Artemis server and NOT from the Github URL!

**Review by kloessst — APPROVED** [review]
re-approve

File: src/main/webapp/app/feature-overview/feature-overview.component.html [hunk 2]
```diff
@@ -0,0 +1,124 @@
+<head>
+    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" />
```
Comment (jpbernius) [inline src/main/webapp/app/feature-overview/feature-overview.component.html:2]: Wait, I think it is better if we host this ourselves! Don't we have FA included already?

File: src/main/webapp/app/feature-overview/feature-overview.scss [hunk 3]
```diff
@@ -0,0 +1,264 @@
+@import url('https://fonts.googleapis.com/css2?family=Ubuntu&display=swap');
```
Comment (jpbernius) [inline src/main/webapp/app/feature-overview/feature-overview.scss:1]: Do we need to load external fonts? What about using the same font setup as elsewhere?

**Review by jpbernius — COMMENTED** [review]
Some Images have the test server badge visible or include Grammarly!

**Comment by FrankeLukas** [comment]
A few style questions:
Imo it would look better if we could center icon and text, if that is possible.
grafik
would become something like
grafik

We should probably think about reducing the text in some of those previews, since it can look very cramped.

Can we use some sort of block-text format? The following is imo a bit hard to read and look at with all the different line lengths:
grafik

Some sort of block-text format with shorter line lengths would also help to better fill the space next to images, since the following has a lot of that:
grafik

File: src/main/webapp/app/feature-overview/feature-overview.component.html (hunk 2 above)
Comment (krusche) [reply src/main/webapp/app/feature-overview/feature-overview.component.html:2]: yes, please use the local version.
It's really `bad practice` to link to external URLs like this

File: src/main/webapp/app/feature-overview/feature-overview.scss (hunk 3 above)
Comment (krusche) [reply src/main/webapp/app/feature-overview/feature-overview.scss:1]: I agree, please use the same font as in all other places.
It's really `bad practice` to link to external URLs like this

**Comment by anditurdiu** [comment]
> A few style questions:
> Imo it would look better if we could center icon and text, if that is possible.
> grafik
> would become something like
> grafik
> 
> We should probably think about reducing the text in some of those previews, since it can look very cramped.
> 
> Can we use some sort of block-text format? The following is imo a bit hard to read and look at with all the different line lengths:
> grafik
> 
> Some sort of block-text format with shorter line lengths would also help to better fill the space next to images, since the following has a lot of that:
> grafik

I wont be able to implement these suggestions in this PR. I like the block text format, maybe that can be done in a follow-up

**Comment by anditurdiu** [comment]
@jpbernius I removed the custom font & the imported icons. Now I this view is using the standard font and the bundled icons. 
@FrankeLukas I implemented the requested changes on the text. I also updated them on the confluence pages:

- https://confluence.ase.in.tum.de/display/ArTEMiS/Feature+Overview+German
- https://confluence.ase.in.tum.de/display/ArTEMiS/Feature+Overview+English

File: src/main/webapp/app/feature-overview/feature-overview.component.html (hunk 2 above)
Comment (anditurdiu) [reply src/main/webapp/app/feature-overview/feature-overview.component.html]: see 6d76e007b13b0e7ef7f2599e75802b607d7d5078 & 00b23dde1932cd3def02faff1e6ac637e80c86bc

File: src/main/webapp/app/feature-overview/feature-overview.scss (hunk 3 above)
Comment (anditurdiu) [reply src/main/webapp/app/feature-overview/feature-overview.scss]: see 6d76e00

File: src/main/webapp/i18n/en/featureOverview.json [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
...questions. Before releasing the exam to your students you can conduct tests runs to make sure everything will run smoothly. Students are limited to a definable amount of time to complete their exam and submit. They can navigate between the exercise and solve them in any order they like.\n\nNo stable internet connection? Not an issue. All changes made by the students are saved locally and automatically synced every 30 seconds.\n\nOnce the exam conduction is over, it is time to assess the submissions. The integrated assessment is done through the tutor exam dashboard which aggregates the progress of all participants to provide a clear overview.\n\nWith all exams assessed you can release the results and allow all students to view their result. If they believe they found a mistake in their assessment they can use the complaint feature to request a re-review. A third assessor, other than the original assessor will then be able to re-assess the specific submission and decide whether to accept or reject the complaint.\n\nFinally, Artemis also creates a statistical report of the exam to evaluate and examine the performance of your students.\n\nCreate, conduct & assess, all within Artemis!"
```
[the remark this answers is not in the document]
Comment (anditurdiu) [reply src/main/webapp/i18n/en/featureOverview.json]: see 6d76e00

File: src/main/webapp/i18n/en/featureOverview.json [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                "configuration": {
+                    "title": "Fully Configurable",
+                    "shortDescription": "Bulk register students from .csv. Configure the exam to your needs. Edit working time of individual students.",
+                    "descriptionTextOne": "Artemis allows you to configure all aspects of the exam. You can change the name of the exam and set the module number, examiner and course name. Furthermore, You can preset the timeline of the exam to fit your needs by deciding when the exam will become visible to the students, when the working time begins and ends and when the results are published. Additionally, you can set the period in which the students can review their results and complain if necessary.\n\nAdd an unlimited amount of exercises from most Artemis exercise types. You can also set exercises to be non-mandatory and add multiple variants to ensure that every student has a unique exam.\n\nFor your convenience, we also included the opportunity to register students in bulk using .csv import or select them from all users registered on Artemis. Additionally, you can also tweak individual student exams for students who require more time."
```
[the remark this answers is not in the document]
Comment (anditurdiu) [reply src/main/webapp/i18n/en/featureOverview.json]: See 6d76e00

File: src/main/webapp/i18n/en/featureOverview.json [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                    "title": "Multiple Exercise Types",
+                    "shortDescription": "Add a variety of exercises, like programming, modeling, text and multiple-choice questions.",
+                    "descriptionTextOne": "Most Artemis exercise types are supported within the exam mode. These include programming-, modeling-, text- and quiz exercises such as multiple choice, drag and drop and short answer questions. You can an unlimited number of exercises to an exam.\n\nFor programming exercises, students decide between working with the integrated code editor or can clone the repository and work on the exercise on their IDE of choice. The modeling editor for modeling exercises is powered by the Apollon editor. You can choose between an array of modeling tasks including class-, activity-, object-, use case-, communication-, component-, deployment diagrams, Petri nets and syntax trees. Your students can answer these questions by intuitively dragging and dropping objects into the editor and connecting them using the appropriate arrows. Through the available variety of exercise types, Artemis fills the requirements of most educational courses."
```
[the remark this answers is not in the document]
Comment (anditurdiu) [reply src/main/webapp/i18n/en/featureOverview.json]: See 6d76e00

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]
Looks very good, here are the last few errors I found:

- Students (EN):
   1. Exam Summary, at the end 
       > Artemis offers **to** option [...]
- Students (DE)
   1. Offline Weiterarbeiten
      > Für Programmieraufgaben **,** brauchst du nur [...]
   2. Qualität und Fairness der Bewertung
       > Diese automatische Bewertung kann durch **einer** manuellen Bewertung [...]
       
       and
       > Diese Funktion kann für Text- und **Modellierübungen** [...]
       
       we used "Modellierungsaufgaben" everywhere else.
       In general, there are a few places where we use "übungen", we should replace all with "aufgaben"
- Instructors (EN)
   1. Conduct Test Runs, at the end
       > Test runs are also a great way to **familiarise** [...]
   2. Automatic Assessment, at the end
       > It then suggests how elements of the submissions should be evaluated **during the course of further correction**.
       
       This is very strange/wordy, it would shorten that to "during further correction" or similar.
   3. Review Individual Exams
       > [...] as an instructor you have **to** ability to access [...]
       
       and
       > You can use this feature to review **indivdual** student exams [...]
- Instructors (DE)
   1. Klausur Erstellen, Durchführen und Bewerten
       > [...] einschließlich Programmier-, Modellierungs-, Text- und Quiz-**Übungen** [...]
   2. Automatische Korrektur, near the end
       > Hier lernt Artemis mittels Maschinelles Lernen [...]
       > Hier lernt Artemis mittels maschinellem Lernens [...]

**Comment by anditurdiu** [comment]
> 1. This is the only occurrence I found here.

See 21aea42692493096128337cd3f45e0ce61905236

**Review by MaisiKoleni — APPROVED** [review]
Looks good to me now. There are more points that could be improved (the German is still a bit strange at some places), but I think we managed to remove almost all errors now.

**Review by kloessst — APPROVED** [review]

**Review by krusche — APPROVED** [review]

**Pull request merged**